extensions-core/datasketches/src/main/java/io/druid/query/aggregation/datasketches/quantiles/QuantilesSketchAggregatorFactory.java

```diff
@@ -0,0 +1,327 @@
+      @JsonProperty("fieldName") String fieldName,
+      @JsonProperty("size") Integer size,
+      @JsonProperty("shouldFinalize") Boolean shouldFinalize,
+      @JsonProperty("isInputSketch") boolean isInputSketch
+  )
+  {
+    this.name = Preconditions.checkNotNull(name, "Must have a valid, non-null aggregator name");
+    this.fieldName = Preconditions.checkNotNull(fieldName, "Must have a valid, non-null fieldName");
+
+    this.size = size == null ? DEFAULT_MAX_SKETCH_SIZE : size;
+
+    //validate the size with sketch library
+    Util.checkIfPowerOf2(this.size, "size");
+    //dummy sketch creation just to ensure all size restrictions are fulfilled.
+    QuantilesSketchUtils.buildSketch(this.size);
+
+    this.shouldFinalize = (shouldFinalize == null) ? true : shouldFinalize.booleanValue();
+    this.isInputSketch = isInputSketch;
+  }
+
+  @SuppressWarnings("unchecked")
+  @Override
+  public Aggregator factorize(ColumnSelectorFactory metricFactory)
+  {
+    ObjectColumnSelector selector = metricFactory.makeObjectColumnSelector(fieldName);
+    if (selector == null) {
+      return new EmptyQuantilesSketchAggregator(name);
+    } else {
+      return new QuantilesSketchAggregator(name, selector, size);
+    }
+  }
+
+  @SuppressWarnings("unchecked")
+  @Override
+  public BufferAggregator factorizeBuffered(ColumnSelectorFactory metricFactory)
+  {
+    ObjectColumnSelector selector = metricFactory.makeObjectColumnSelector(fieldName);
+    if (selector == null) {
+      return new EmptyQuantilesSketchBufferAggregator();
+    } else {
+      return new QuantilesSketchBufferAggregator(selector, size, getMaxIntermediateSize());
+    }
+  }
+
+  @Override
+  public Object deserialize(Object object)
+  {
+    return QuantilesSketchUtils.deserialize(object);
+  }
+
+  @Override
+  public Comparator<QuantilesSketch> getComparator()
+  {
+    return COMPARATOR;
+  }
+
+  @Override
+  public Object combine(Object lhs, Object rhs)
+  {
+    if (lhs == null || ((QuantilesSketch) lhs).isEmpty()) {
+      return rhs;
+    }
+
+    if (rhs == null || ((QuantilesSketch) rhs).isEmpty()) {
+      return lhs;
+    }
+
+    Union union = QuantilesSketchUtils.buildUnion(size);
+    union.update((QuantilesSketch) lhs);
+    union.update((QuantilesSketch) rhs);
+
+    return union.getResult();
+  }
+
+  @Override
+  @JsonProperty
+  public String getName()
+  {
+    return name;
+  }
+
+  @JsonProperty
+  public String getFieldName()
+  {
+    return fieldName;
+  }
+
+  @JsonProperty
+  public int getSize()
+  {
+    return size;
+  }
+
+  @JsonProperty
+  public boolean getShouldFinalize()
+  {
+    return shouldFinalize;
+  }
+
+  @JsonProperty
+  public boolean getIsInputSketch()
+  {
+    return isInputSketch;
+  }
+
+  @Override
+  public int getMaxIntermediateSize()
+  {
+    // off-heap implementation of quantiles sketch is not available right now and hence
+    // nothing is stored off-heap really.
+    return 1;
```

Comment:
ok, let us do (1) optionally , let user specify maxOffheapSize to use.. if sketch fits within that size then it gets stored off-heap (with significant additional serde cost) or else it gets stored on-heap. Now users can decide where they want to get it stored.
we should still keep #2827 so that user can choose to store on-heap but tune topN if/when necessary.